nova_vm/src/ecmascript/builtins/typed_array/data.rs

```diff
@@ -3,13 +3,168 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use crate::{
-    ecmascript::types::OrdinaryObject,
+    ecmascript::{builtins::ArrayBuffer, types::OrdinaryObject},
     heap::{CompactionLists, HeapMarkAndSweep, WorkQueues},
 };
 
-#[derive(Debug, Clone, Default)]
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct TypedArrayByteLength(pub u32);
+
+impl TypedArrayByteLength {
+    pub const fn value(value: u32) -> Self {
+        Self(value)
+    }
+
+    /// A sentinel value of `u32::MAX - 1` means that the byte length is stored in an
+    /// associated map in the heap. This will most likely be a very rare case,
+    /// only applicable for 4GB+ buffers.
+    pub const fn heap() -> Self {
+        Self(u32::MAX - 1)
+    }
+
+    /// A sentinel value of `u32::MAX` means that the byte length is the
+    /// `AUTO` value used in the spec.
+    pub const fn auto() -> Self {
+        Self(u32::MAX)
+    }
+}
+
+impl Default for TypedArrayByteLength {
+    fn default() -> Self {
+        Self::auto()
+    }
+}
+
+impl From<Option<usize>> for TypedArrayByteLength {
+    fn from(value: Option<usize>) -> Self {
+        match value {
+            Some(value) => {
+                if value >= Self::heap().0 as usize {
+                    Self::heap()
+                } else {
+                    Self::value(value as u32)
+                }
+            }
+            None => Self::auto(),
+        }
+    }
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct TypedArrayArrayLength(pub u32);
+
+impl TypedArrayArrayLength {
+    pub const fn value(value: u32) -> Self {
+        Self(value)
+    }
+
+    /// A sentinel value of `u32::MAX - 1` means that the byte length is stored in an
+    /// associated map in the heap. This will most likely be a very rare case,
+    /// only applicable for 4GB+ buffers.
+    pub const fn heap() -> Self {
+        Self(u32::MAX - 1)
+    }
+
+    /// A sentinel value of `u32::MAX` means that the byte length is the
+    /// `AUTO` value used in the spec.
+    pub const fn auto() -> Self {
+        Self(u32::MAX)
+    }
+}
+
+impl Default for TypedArrayArrayLength {
+    fn default() -> Self {
+        Self::auto()
+    }
+}
+
+impl From<Option<usize>> for TypedArrayArrayLength {
+    fn from(value: Option<usize>) -> Self {
+        match value {
+            Some(value) => {
+                if value >= Self::heap().0 as usize {
+                    Self::heap()
+                } else {
+                    Self::value(value as u32)
+                }
+            }
+            None => Self::auto(),
+        }
+    }
+}
+
+// TODO: Investigate if the common case is that the byte offset is less than
+// an u16, that would mean we could squeeze an extra 2 bytes out of the struct.
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct TypedArrayByteOffset(pub u32);
```

Comment:
**question:** Could we reuse the DataView byte offset and length structs? We could call then `ViewedArrayBufferByteLength` and `-Offset` if we want to.

Reply:
Sure!